modin/core/dataframe/pandas/dataframe/utils.py

```diff
@@ -517,3 +517,212 @@ def run_f_on_minimally_updated_metadata(self, *args, **kwargs):
+        complete_index = pandas.MultiIndex.from_product(
+            [
+                value.categories.astype(total_level.dtype)
+                for total_level, value in zip(
+                    total_index.levels, missing_cats_dtype.values()
+                )
+            ],
+            names=by,
+        )
+        missing_index = complete_index[~complete_index.isin(total_index)]
+    else:
+        if not isinstance(total_index, pandas.CategoricalIndex):
+            return {}, new_combined_cols
+        # if we're grouping on a single grouper then we simply compute the difference
+        # between categorical values in the result and the values defined in categorical dtype
+        missing_index = total_index.categories.difference(total_index.values)
+        missing_cats_dtype = {by[0]: pandas.CategoricalDtype(missing_index)}
+    missing_index.names = by
+
+    if len(missing_index) == 0:
+        return {}, new_combined_cols
+
+    ### At this stage we want to get a fill_value for missing categorical values
+    if is_udf_agg and isinstance(total_index, pandas.MultiIndex):
+        # if grouping on multiple columns and aggregating with an UDF, then the
+        # fill value is always `np.NaN`
+        missing_values = pandas.DataFrame({0: [np.NaN]})
+    else:
+        # In case of a udf aggregation we're forced to run the operator against each
+        # missing category, as in theory it can return different results for each
+        # empty group. In other cases it's enough to run the operator against a single
+        # missing categorical and then broadcast the fill value to each missing value
+        if not is_udf_agg:
+            missing_cats_dtype = {
+                key: pandas.CategoricalDtype(value.categories[:1])
+                for key, value in missing_cats_dtype.items()
+            }
+
+        empty_df = pandas.DataFrame(columns=initial_columns)
+        # HACK: default 'object' dtype doesn't fit our needs, as most of the aggregations
+        # fail on a non-numeric columns, ideally, we need dtypes of the original dataframe,
+        # however, 'int64' also works fine here if the original schema is not available
+        empty_df = empty_df.astype(
+            "int64" if initial_dtypes is None else initial_dtypes
+        )
+        empty_df = empty_df.astype(missing_cats_dtype)
+        missing_values = operator(empty_df.groupby(by, **kwargs))
+
+    if is_udf_agg and not isinstance(total_index, pandas.MultiIndex):
+        missing_values = missing_values.drop(columns=by, errors="ignore")
+        new_combined_cols = pandas.concat(
+            [
+                pandas.DataFrame(columns=combined_cols),
+                missing_values.iloc[:0],
+            ],
+            axis=0,
+            join="outer",
+        ).columns
+    else:
+        # If the aggregation has failed, the result would be empty. Assuming the
+        # fill value to be `np.NaN` here (this may not always be correct!!!)
```

Comment:
So is this also a hack?

Reply:
yes, this is a hack, added an according note